Add `Redis` Session driver

circulon opened the pull request:
This is an initial implementation of a driver for Redis backed sessions

@josephmancuso @girardinsamuel 
Interested in your opinions on how I have implemented this
 
Tests are not built yet.... coming soon

circulon committed:
Added a Redis driver for session usage

circulon committed:
Cookie dirver inherits from BaseDriver

circulon committed:
Handle MessageBag validation error messages

girardinsamuel reviewed (commented):
For now the architecture looks good !

girardinsamuel commented:
Hi ! 
Looks like linting needs to be fixed on this PR 😉

circulon committed:
Fixed linting issues

circulon committed:
Added session drivers find and load their own config

circulon committed:
Added Redis Driver defaults

circulon committed:
Aligned method names for compatibility

circulon committed:
Fixed default driver always used even if named driver specified

circulon committed:
Explicitly name the driver being tested

circulon committed:
Removed duplicate test

circulon committed:
Added Redis Session Driver test

circulon commented:
@girardinsamuel @josephmancuso 

OK I think this is completed now.
Added tests and reworked config and general compatibility with rxisting session methods

circulon committed:
Fixed driver not found if using default

circulon committed:
Added a test to make sure the default session driver is configured

circulon committed:
check default session driver is defined in config

circulon committed:
Moved the check for default driver and its config for initial setup checking

josephmancuso commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -0,0 +1,147 @@
+        if not session_id:
+            request = self.application.make("request")
+            session_id = request.cookie("SESSID")
+        ns = self.options.get("namespace", "")
```
ns variable should be called `namespace`. Remember code is read more than its written. No need for shorthanded variable names

circulon replied:
Fixed

josephmancuso commented on src/masonite/sessions/Session.py:
```diff
@@ -31,19 +34,31 @@ def driver(self, driver: str) -> Any:
     def set_configuration(self, config: dict) -> "Session":
         """Set session driver options."""
+
+        # make sure the default drivewr is defined
```
typo. `drivewr` -> `driver`

circulon replied:
Fixed

josephmancuso commented on tests/integrations/config/session.py:
```diff
@@ -4,4 +4,13 @@
     "default": "cookie",
     "cookie": {},
+    "redis": {
+        "driver_config": {
```
Why are we wrapping this inside a driver_config dictionary? Can' we just do `host`, `port`, `password` inside the `redis` connection? This looks ugly to me and not really similar to other features

circulon replied:
The reason for this is that if you do this in the same way as per the Redis Cache Driver there is no easy way to provide additional initial configuration to the underlying connection driver StrictRedis.
 ie how do I set the "db=3" on the connection or enable any of the ssl options?
There is no facility for this at all nor on any of the other drivers that I could tell 
In the case of the StrictRedis driver there are 35+ options available to configure the driver.

This  might cause the developer for example to write a subclass of the RedisDriver to override the get_connection() method to use out of the box  options that the driver module provides. 

Yes it is a bit ugly I admit but I could not see a better way to describe the options that would be applied to the module driver itself.
While still providing configuration to the wrapper (session in this case) for things like timeout and namespace.  
Putting all module config and wrapper config options at the same level means you then have to separate them at the wrapper level by some other means.   

Ideally the framework should provide the ability to use the underlying module the ServiceProvider wraps with any configuration that is defined by the underlying module/s it uses in as easy a manner as possible. 
Not force the developer  to write additional code to handle something the underlying module (driver) already provides.

I hope this clarifies the design choice.

Happy to discuss more about better ways to provide this functionality

girardinsamuel replied:
Maybe the solution is to do something like this:

```python
DRIVERS = {
    #...
    "redis": {
        "host": "127.0.0.1",
        "port": "6379",
        "password": "",
        "options": {},
        "namespace": "",
        "timeout": "",
    }
}
```
and then build the connection like this:

```python
self.connection = Redis(
    **self.options.get("options"),
    host=self.options.get("host"),
    port=self.options.get("port"),
    password=self.options.get("password"),
)
```

(It's what is used for Masonite ORM config)

circulon replied:
@girardinsamuel  thanks thats much cleaner and still provides flexibility.
Implemented it and its working ;)

josephmancuso reviewed (changes requested):
Looking good! I gave it a code review. Just some things I saw and some comments to make the code more similar to our other features

circulon committed:
Fixed Typo

circulon committed:
Expanded variable names for better readability

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -0,0 +1,147 @@
+            )
+
+        if not self.connection:
+            self.connection = redis.StrictRedis(
```
On an other note maybe we could upgrade redis version and use Redis (https://github.com/redis/redis-py#client-classes-redis-and-strictredis) since 3.X StrictRedis has been renamed to Redis.

circulon replied:
Yeah thats makes sense and keeps Masonite up to date with the backing module 

> On an other note maybe we could upgrade redis version and use Redis (https://github.com/redis/redis-py#client-classes-redis-and-strictredis) since 3.X StrictRedis has been renamed to Redis.

That work for me and keeps Masonite inline with the backend module

girardinsamuel commented on tests/features/session/test_cookie_session.py:
```diff
@@ -6,42 +6,35 @@ def test_can_start_session(self):
         session = self.application.make("session")
         request.cookie("s_hello", "test")
-        session.start()
+        session.start("cookie")
```
Can you explain this change ? 🤔 I am not sure to follow.

circulon replied:
In this case although the "default" config for these tests is defined as "cookie" this may be changed by someone else when building other tests for other drivers.
This makes sure we are explicitly testing the "cookie" driver and not relying on the "default" which may or may not be "cookie"

circulon replied:
Re: the driver config ... I'm still learning the nuances of Python re passing options that have not actually assignments for the class.  
ie I assumed (incorrectly) that the class init would complain that the "timeout" and "namespace" kwargs which are not actually Redis() (or StrictRedis()) args would not work.

so yeah if this works I'm fine with it and so it will align with the other configs and reduce complexity.  

Will report back if this doesnt work as expected.

circulon replied:
@girardinsamuel 
after rereading your "options" config again this make much more sense... sorry its been a long day and I'm tired ;)  

so Yes the "options" at the root of the driver config makes much more sense and is a much cleaner option...
Will test this scenario in the next couple of days or so.  
Cheers for the pointers

circulon committed:
Moved underlying module (driver) config to the root level

circulon committed:
Make sure the keys are strings before searching them

circulon committed:
Dont rely on “default” driver as it may change

circulon committed:
Adjusted test config

circulon committed:
Added additional Redis options for testing

circulon committed:
Fixed driver default options not set correctly

girardinsamuel commented:
Hi @circulon, thank you again for your contribution.
One thing I missed in the first review are the type hints. We are progressively type hinting the whole Masonite codebase (It should be completed for Masonite 5). As the Cookie session driver is type hinted, could you do the same for Redis session driver ? You can check how it's done in `masonite/drivers/session/CookieDriver.py`. 👍 

Note: i have seen it's partially done, so it should just be adding some missing type hints on some methods.

circulon committed:
Added connection method hinting

circulon committed:
Added more type hinting

circulon commented:
> Hi @circulon, thank you again for your contribution. One thing I missed in the first review are the type hints. We are progressively type hinting the whole Masonite codebase (It should be completed for Masonite 5). As the Cookie session driver is type hinted, could you do the same for Redis session driver ? You can check how it's done in `masonite/drivers/session/CookieDriver.py`. 👍
> 
> Note: i have seen it's partially done, so it should just be adding some missing type hints on some methods.

Hi @girardinsamuel 
I appreciate the thanks ;)

Yeah my project requires Redis for the cache and I use a thitf party auth system so internal lookups to validate the user token are just not feasible.
So I decided to put the user object in the session as many other frameworks do.  Turns out having tokens in the user object exceeds the cookie size limit ;(  

As the Redis Cache driver was already built it was just a matter of reworking that driver to fit the way Sessions work.
 So there is a lot of reused code from the cache driver as you might have noticed.

Regarding Type hinting I have added most of them.  There are a couple of methods like "set" that I'm not exactly sure of the return type so any pointers would be appreciated.

girardinsamuel commented on tests/integrations/config/session.py:
```diff
@@ -5,11 +5,8 @@
-            "port": "6379",
-            "password": "",
-        },
+        "host": "127.0.0.1",
```
I know you put default values inside Redis session driver connection, but could you in addition put "port" and "password" here again ? It will help users wanting to use this driver to configure it as they will directly see those base options. 👍 Also it will more consistent with options of redis cache driver.

circulon replied:
Yeah Makes sense...
Done

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -81,25 +86,25 @@ def save(
             self.delete(f"{self.get_flash_prefix()}{key}")
 
-    def get_session_namespace(self, session_id=None):
+    def get_session_namespace(self, session_id=None) -> str:
```
```suggestion
    def get_session_namespace(self, session_id: str = None) -> str:
```

circulon replied:
Aah yes I see what you mean ok ....  Done

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -81,25 +86,25 @@ def save(
         return self
 
-    def get(self, key, default=None, **options):
+    def get(self, key, default=None, **options) -> str:
```
```suggestion
    def get(self, key: str, default: Any = None, **options) -> str:
```

girardinsamuel replied:
For set() you can type hint like this:
```python
def set(self, key: str, value: Any, time: int = None):
```

For has() you can type hint like this:
```python
def has(self, key: str) -> bool:
```

circulon replied:
Yeah I see ... For some reason I thought "type hinting" meant the hint for the return value of methods.
I learnt something new today... 
Cheers

Done

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -1,6 +1,8 @@
 import json
 from typing import TYPE_CHECKING
```
```suggestion
from typing import TYPE_CHECKING, Any
```

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -120,9 +125,9 @@ def has(self, key):
     def delete(self, key):
```
```suggestion
    def delete(self, key: str):
```

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -120,9 +125,9 @@ def has(self, key):
         return self.get_connection().delete(key)
 
-    def flush(self, session_id=None):
+    def flush(self, session_id=None) -> None:
```
```suggestion
    def flush(self, session_id: str = None) -> None:
```

girardinsamuel commented on src/masonite/drivers/session/RedisDriver.py:
```diff
@@ -133,11 +138,11 @@ def flush(self, session_id=None):
+        return int(self.options.get("timeout", 60 * 60 * 24))
 
-    def get_value(self, value):
+    def get_value(self, value) -> str:
```
```suggestion
    def get_value(self, value: Any) -> str:
```

girardinsamuel reviewed (commented):
Here are missing type hints to help you.
For some return types you have to look at redis methods return type to know what is the type.

circulon committed:
Readded port and password options for consistency

circulon committed:
Added type hingint for parameters in RedisDriver

circulon committed:
Added missing return value hint

girardinsamuel reviewed (approved):
Final review ! Looks good to me.
@josephmancuso I guess we can merge this now.

josephmancuso reviewed (approved):
Final code review looks good
